ai_spider/stats.py

```diff
@@ -6,30 +6,176 @@
 from threading import Thread
 
 from ai_spider.unique_queue import UniqueQueue
+from sqlalchemy import and_
 
 STATS_EMA_ALPHA = 0.9
 PUNISH_SECS = 60 * 15
 PUNISH_BAD_PERF = 9999
 
+class StatsStore:
+    def __init__(self, session=None):
+        self.update_queue = UniqueQueue(key=lambda el: el[0])
+        self.update_thread = Thread(daemon=True, target=self.update_loop)
+        self.update_thread.start()
+        self.session=session
+
+    def stop(self, join=True):
+        self.update_queue.put([None, None])
+        if join:
+            self.update_thread.join()
+
+    def update_loop(self):
+        while True:
+            try:
+                key, bnum, val = self.update_queue.get(timeout=4)
+                if key is None:
+                    break
+                self._update(key, bnum, val)
+            except Empty:
+                pass
+
+    def update(self, key: str, bin_num: int,  val: dict):
+        self.update_queue.put([key, bin_num, val])
+
+    def _update(self, key: str, bin_num: int, vals: dict):
+        raise NotImplementedError
+
+    def get(self, key: str, bin_num: int, filters: dict) -> dict:
```

Comment:
no bin_num needed

Reply:
I was following how you laid it out here where bin_num was being passed in: https://github.com/ArcadeLabsInc/queenbee/pull/9#issuecomment-1760019361